### ISSUE TITLE
Go API should return errors instead of panicking 

### ISSUE THREAD

**Issue by cmars** [attic-labs/noms#2367]
Currently errors are not propagated up to the caller in the Go API. If an operation fails, in many cases `d.Chk.NoError` or `d.PanicOnError` is called to give up and unwind the stack.

These kind of assertions are fine for test code, but problematic for production code that would use noms as a library. To an extent, panics can be co-opted into a "try-catch" mechanism, but in Go, panics are not really intended to be used as exceptions. `panic` is typically reserved for cases where a goroutine's situation is fubar to the point where the situation cannot be recovered.

There might be some noms failure modes where the caller would agree that it's panic-time, but by panicking, you're forcing that decision onto every API consumer. The caller should be able to decide how to handle the errors, especially since these errors may be due to externalities like the file system, network, etc. and you can't really predict every intended use by your API consumers.

It's also going to be difficult to maintain noms over time with this approach. Case in point: comments in [datas.closeResponse](https://github.com/attic-labs/noms/blob/cabc7aeb8caa4d6482a250d519c56d6f0e6d90dc/go/datas/http_batch_store.go#L507), it looks like you're already having trouble with shadowed errors. If errors are wrapped with stack context at the point of detection and returned, they're going to be much more observable.

That said, reading a bit into the code base, I can understand why the panicking approach was taken. It's pretty difficult to wind errors through worker queues when you're using `WaitGroup`s and channels to coordinate workers. The panic approach was probably the best way to get things working quickly, so all in all, perhaps a good decision to get the project moving!

Some suggestions:
- Every function / method that can fail should return an error.
- Wrap errors with stack tracing. I like `github.com/juju/errors` for this because it also has some nice error types, but there are several packages that do this.
- Use [tombs](http://blog.labix.org/2011/10/09/death-of-goroutines-under-control) (`gopkg.in/tomb.v2` is the latest version) to manage workers, replacing the bespoke wait group / channel termination. This should simplify worker code and helps get errors out of the workers where the fail-able stuff happens (HTTP requests and what not).
- Eliminate the `MaybeXYZ` methods -- just `XYZ`. If the target (head, head ref, etc.) is not found, then return a "not found" error type. For example, `github.com/juju/errors.NewNotFound("dataset has no head / ref / etc")`. You could then use errors.IsNotFound(err) -- possibly adding a `HasXYZ(...) (bool, error)` method if there'd be too much boilerplate around doing that everywhere.

Admittedly, I've barely scratched the surface of the noms codebase. I'm just playing with these ideas in a branch to see how well it works out -- but who knows, there might be more dragons lurking with regard to errors. Before I go too much farther, I'd like to hear some suggestions and feedback.

Thanks!
Casey

**Comment by aboodman** [comment]
Hi Casey,

Thanks for getting involved -- I've looked through your github repos and I'd love for you to contribute to Noms.

That said, you picked a thorny area to start in :) We have spent a _lot_ of time debating error handling in Go.

Our original motivation for using panics was simply the common Go complaint that checking errors on every stack frame and manually propagating them is tedious, obscures the intent of the code, makes it easy to accidentally swallow errors, and makes maintenance tedious. This is something that even the core Go team has run into and [resolved](https://golang.org/src/encoding/json/decode.go#L152) the same way we did.

That said, I think there is still room for improvement. For example, I'd personally like panics to be less a part of our _external_ API than they currently are.

I think the most productive way to help is to be specific. It's so easy to get philosophical about error handling -- 'always do x', 'never do y', etc. The Noms codebase is big enough that such pronouncements will always find counterexamples.

Instead, if you pick one specific method and say: "it would be easier for me to use noms if this method has a different api because of reasons: a, b, and c", then it will be much easier for us to have a productive (and hopefully quick) conversation.

In particular if your proposal is accompanied by a working, small, patch that can be debated, then it's even easier to have a discussion.

Thanks, and happy hacking!
